### ISSUE TITLE
AVR Mining with Raspberry Pi Pico

### ISSUE THREAD

**Issue by just-daviq** [duino-coin/duino-coin#996]
![Snímka](https://user-images.githubusercontent.com/68821513/132995406-10b29c4d-ba0b-4935-a2f2-d65a95817289.PNG)

I mined for 5 hours and I still had this, it didn't work out

**Comment by gsrbr** [comment]
Hello @Daviidkoo4074 what AVR are you using?

**Comment by just-daviq** [comment]
Raspberry Pi Pico

**Comment by gsrbr** [comment]
This AVR is not in a very delightful position, in the latest version of the avr miner it(PICO) is not functional... We are still studying its possible reactivation

**Comment by just-daviq** [comment]
So could I upgrade to lower versions?

**Comment by revoxhere** [comment]
Hello, to mine with Raspberry Pi pico you need to set a higher avr timeout value in the Settings.cfg file

**Comment by Izder456** [comment]
> 
> 
> Hello, to mine with Raspberry Pi pico you need to set a higher avr timeout value in the Settings.cfg file

whats an example value? i have 45 set in my config and nothing is fixed.

**Comment by revoxhere** [comment]
Hello, @gsrbr informed me that the Pico mining is not working properly because of the unique id issue. You will need to wait until one of our AVR devs pushes a fix to this

**Comment by Izder456** [comment]
> 
> 
> Hello, @gsrbr informed me that the Pico mining is not working properly because of the unique id issue. You will need to wait until one of our AVR devs pushes a fix to this

ok thanks
if you need to get a pico, don't get it on amazon. the price is doubled on there $8USD versus the Manufacturer recommended $4USD.

**Comment by revoxhere** [comment]
Thanks for the tip!

**Comment by raymate** [comment]
You can run it on a Pico, anyone got instructions?

**Comment by gsrbr** [comment]
Hello @raymate as said by revox and me, the Pico mining is not working properly because of the unique id issue. You will need to wait until one of our AVR devs pushes a fix to this

**Comment by Izder456** [comment]
I dont mean to be a "wen ETA" kid, but is there plans to support rpi pico in the future?

**Comment by primitt** [comment]
Its hard to support every single board, but avr code should work on rpi pico (when devs get around to fixing it)

**Comment by JK-Rolling** [comment]
These are the steps I performed to get RPi Pico mining.

1. open up IDE, Tools --> Board --> Board Manager --> Search `pico`
2. Look for `Arduino Mbed OS RP2040 Boards` --> Select version `2.3.1` --> Install
3. Optional - Close and reopen IDE
4. Open `Arduino_Code` sketch
5. Choose the right board. Tools --> Board --> Arduino Mbed OS RP2040 Boards --> Raspberry Pi Pico
6. Click Compile (there shouldn't be any error)
5. For new Pico, press and hold `BOOTSEL` while insert USB, it should be detected as external drive (there shouldn't be any COM port selected in IDE)
6. Click upload
7. Unplug and re-plug in the USB (without holding BOOTSEL)
8. You may start the Python miner to start mining

**Comment by Izder456** [comment]
> These are the steps I performed to get RPi Pico mining.
> 
> 
> 
> 1. open up IDE, Tools --> Board --> Board Manager --> Search `pico`
> 
> 2. Look for `Arduino Mbed OS RP2040 Boards` --> Select version `2.3.1` --> Install
> 
> 3. Optional - Close and reopen IDE
> 
> 4. Open `Arduino_Code` sketch
> 
> 5. Choose the right board. Tools --> Board --> Arduino Mbed OS RP2040 Boards --> Raspberry Pi Pico
> 
> 6. Click Compile (there shouldn't be any error)
> 
> 5. For new Pico, press and hold `BOOTSEL` while insert USB, it should be detected as external drive (there shouldn't be any COM port selected in IDE)
> 
> 6. Click upload
> 
> 7. Unplug and re-plug in the USB (without holding BOOTSEL)
> 
> 8. You may start the Python miner to start mining
> 
> 

Thanks! I'll have a go at it when I get home. Wish me luck!

**Comment by Adytzu4u79** [comment]
> These are the steps I performed to get RPi Pico mining.
> 
> 1. open up IDE, Tools --> Board --> Board Manager --> Search `pico`
> 2. Look for `Arduino Mbed OS RP2040 Boards` --> Select version `2.3.1` --> Install
> 3. Optional - Close and reopen IDE
> 4. Open `Arduino_Code` sketch
> 5. Choose the right board. Tools --> Board --> Arduino Mbed OS RP2040 Boards --> Raspberry Pi Pico
> 6. Click Compile (there shouldn't be any error)
> 7. For new Pico, press and hold `BOOTSEL` while insert USB, it should be detected as external drive (there shouldn't be any COM port selected in IDE)
> 8. Click upload
> 9. Unplug and re-plug in the USB (without holding BOOTSEL)
> 10. You may start the Python miner to start mini

**Comment by Adytzu4u79** [comment]
Hello, it doesn't work for me.

**Comment by Izder456** [comment]
> > These are the steps I performed to get RPi Pico mining.
> > 
> > 1. open up IDE, Tools --> Board --> Board Manager --> Search `pico`
> > 2. Look for `Arduino Mbed OS RP2040 Boards` --> Select version `2.3.1` --> Install
> > 3. Optional - Close and reopen IDE
> > 4. Open `Arduino_Code` sketch
> > 5. Choose the right board. Tools --> Board --> Arduino Mbed OS RP2040 Boards --> Raspberry Pi Pico
> > 6. Click Compile (there shouldn't be any error)
> > 7. For new Pico, press and hold `BOOTSEL` while insert USB, it should be detected as external drive (there shouldn't be any COM port selected in IDE)
> > 8. Click upload
> > 9. Unplug and re-plug in the USB (without holding BOOTSEL)
> > 10. You may start the Python miner to start mining
> 
> Thanks! I'll have a go at it when I get home. Wish me luck!


these steps work as long as you have a linux host that runs the AVR_Miner.py script (such as a rpi 3b+)

> Hello, it doesn't work for me.


try dual booting linux and run the AVR_Miner.py from there. or just get a regular raspberry pi and run it in SCREEN. the only caveot is that the first 10-ish shares get rejected for some reason, even if you up the AVR timeout in Settings.cfg

**Comment by ucra-fear** [comment]
I followed these steps on Raspberry PI OS Buster and Bullseye, as well as Windows 10 and I get the  same error when trying to upload to the pico board.  The error is: 
rp2040load 1.0.1 - compiled with go1.15.8
....................................
An error occurred while uploading the sketch.  

Update:
Wow.  I think I resolved it on Windows. Every time I closed and reopened Arduino IDE, it would tell me there were updates to installed boards and libraries.. so I went back to the Arduino Embedded Board RP2040 and pressed update.. it then updated the 2.3.1 to 2.6.1... I figured what the hell.  It still gave the same error, BUT, the a series of animated dashes (----------) happened with the text loading into Flash 100% and then it stated Done Uploading and then Windows told me device was ready.....

One other thing I should point out -  I chose a 2nd pico but with headers to test and see if the headers was it.. it was not.   BUT on the first pico   about 20 mins after I had been trying Windows popped up asking it I wanted to allow a file to modify my PC.  It looked to be something AMD related, but they were NOT signed,  so I told it Do not allow.

Well when I plugged int he 2nd Pico that same thing popped up right away. Still unassigned and looked to be related to AMD. So I took a gamble and it was additional USB Serial Drivers from Arduino.  -  This may be why it worked on Windows.

I am trying the Board update on Raspberry Pi now to see what happens.
Update 2 -The board update option did not fix the issue on Raspberry Pi.  So it must have been that unsigned USB Serial Driver from Arduino that fixed it on Windows.

So What I am doing for now is program the Pico on my one remaining Widows box, then plug it to the Raspberry Pi and it is detected as a Raspberry Pi Pico.

**Comment by ucra-fear** [comment]
This is a continuation of the above testing I did.  I mentioned the info below in an open Issue I opened 2nd part of #1121. 

I tried using the AVR_Miner.py on a Raspberry Pi 4 running 3 Raspberry Pi Picos and got a similar or same issue.
It connected fine once but then didn't see the pico so I re-flashed the pico and the miner detected all 3 as Raspberry PI Pico.

Now when I try to run the AVR_Miner I get the following output right away.
pi@RP400-8:~/duino-coin $ python3 AVR_Miner.py

00:20:28 net0 Searching for the fastest node to connect to
00:20:30 sys0 AVR mining thread is starting using DUCO-S1A algorithm (/dev/ttyACM0)
00:20:30 net0 Connected to master Duino-Coin server (v2.7)
00:20:30 sys1 AVR mining thread is starting using DUCO-S1A algorithm (/dev/ttyACM1)
00:20:31 net0 MOTD: You are mining on Bilapool, have fun!
00:20:31 sys2 AVR mining thread is starting using DUCO-S1A algorithm (/dev/ttyACM2)
Exception in thread Thread-3:
Traceback (most recent call last):
File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
self.run()
File "/usr/lib/python3.9/threading.py", line 892, in run
self._target(*self._args, **self._kwargs)
File "/home/pi/duino-coin/AVR_Miner.py", line 872, in mine_avr
share_print(port_num(com), "reject",
File "/home/pi/duino-coin/AVR_Miner.py", line 653, in share_print
print(Fore.WHITE + datetime.now().strftime(Style.DIM + "%H:%M:%S ")
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 41, in write
self.__convertor.write(text)
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 164, in write
self.wrapped.write(text)
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 41, in write
self.__convertor.write(text)
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 164, in write
self.wrapped.write(text)
UnicodeEncodeError: 'latin-1' codec can't encode character '\u26cf' in position 54: ordinal not in range(256)
^C00:20:35 sys0 SIGINT detected - Exiting gracefully. See you soon!

I edited the config file to turn on Debugging and got this output.
pi@RP400-8:~/duino-coin $ python3 AVR_Miner.py
00:15:04.896264 DEBUG: Config file loaded

00:15:04.901157 DEBUG: Error displaying greeting message: 'latin-1' codec can't encode character '\u2016' in position 10: ordinal not in range(256)
00:15:04.901495 DEBUG: Starting donation process
00:15:04.906583 DEBUG: Error launching donation thread: 'latin-1' codec can't encode characters in position 83-84: ordinal not in range(256)
00:15:04 net0 Searching for the fastest node to connect to
00:15:05.326033 DEBUG: Fetched pool: bilapool-2
00:15:07.332352 DEBUG: Connecting to ('51.158.113.59', 6043)
00:15:07.332853 DEBUG: Connecting to ('51.158.113.59', 6043)
00:15:07.333120 DEBUG: Connecting to ('51.158.113.59', 6043)
00:15:07 net0 Connected to master Duino-Coin server (v2.7)
00:15:07 sys1 AVR mining thread is starting using DUCO-S1A algorithm (/dev/ttyACM1)
00:15:07.651268 DEBUG: /dev/ttyACM1: Requesting job
00:15:07 sys0 AVR mining thread is starting using DUCO-S1A algorithm (/dev/ttyACM0)
00:15:07.656853 DEBUG: /dev/ttyACM0: Requesting job
00:15:07 net0 MOTD: You are mining on Bilapool, have fun!
00:15:07 sys2 AVR mining thread is starting using DUCO-S1A algorithm (/dev/ttyACM2)
00:15:07.801592 DEBUG: /dev/ttyACM2: Requesting job
00:15:07.801819 DEBUG: /dev/ttyACM1: Received: e340d8c0d7915915d86bd7d5d43b3c1c85d71256
00:15:07.802646 DEBUG: /dev/ttyACM1: Sending job to the board
00:15:07.801976 DEBUG: /dev/ttyACM0: Received: e340d8c0d7915915d86bd7d5d43b3c1c85d71256
00:15:07.804136 DEBUG: /dev/ttyACM0: Sending job to the board
00:15:07.804868 DEBUG: /dev/ttyACM1: Reading result from the board
00:15:07.805048 DEBUG: /dev/ttyACM0: Reading result from the board
00:15:07.888045 DEBUG: /dev/ttyACM0: Result: 11011110
00:15:07.917504 DEBUG: /dev/ttyACM2: Received: fea555c58970d4dff5d40f033e22796e5cb355cc
00:15:07.917919 DEBUG: /dev/ttyACM2: Sending job to the board
00:15:07.918494 DEBUG: /dev/ttyACM2: Reading result from the board
00:15:07.919464 DEBUG: /dev/ttyACM1: Result: 100110111
00:15:07.973486 DEBUG: /dev/ttyACM2: Result: 10001100
00:15:08.000624 DEBUG: /dev/ttyACM0: retrieved feedback: BAD,Incorrect difficulty
Exception in thread Thread-3:
Traceback (most recent call last):
File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
self.run()
File "/usr/lib/python3.9/threading.py", line 892, in run
self._target(*self._args, **self._kwargs)
File "/home/pi/duino-coin/AVR_Miner.py", line 872, in mine_avr
share_print(port_num(com), "reject",
File "/home/pi/duino-coin/AVR_Miner.py", line 653, in share_print
print(Fore.WHITE + datetime.now().strftime(Style.DIM + "%H:%M:%S ")
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 41, in write
self.__convertor.write(text)
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 164, in write
self.wrapped.write(text)
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 41, in write
self.__convertor.write(text)
File "/usr/lib/python3/dist-packages/colorama/ansitowin32.py", line 164, in write
self.wrapped.write(text)
UnicodeEncodeError: 'latin-1' codec can't encode character '\u26cf' in position 54: ordinal not in range(256)
00:15:08.043052 DEBUG: /dev/ttyACM1: retrieved feedback: BAD,Incorrect difficulty
00:15:08.086603 DEBUG: /dev/ttyACM2: retrieved feedback: BAD,Incorrect difficulty
